storage/wdpost/wdpost_run_faults.go

```diff
@@ -343,3 +344,142 @@ func (s *WindowPoStScheduler) asyncFaultRecover(di dline.Info, ts *types.TipSet)
+		}
+		ptxinfo := ptx{
+			deadline:  ptxID.Deadline,
+			partition: ptxID.Partition,
+		}
+
+		slist := smap[ptxinfo]
+		sn := uint64(sector)
+		slist = append(slist, sn)
+		smap[ptxinfo] = slist
+	}
+
+	for i, v := range smap {
+		sectorinbit := bitfield.NewFromSet(v)
+		RecoveryDecls = append(RecoveryDecls, miner.RecoveryDeclaration{
+			Deadline:  i.deadline,
+			Partition: i.partition,
+			Sectors:   sectorinbit,
+		})
+	}
+
+	// Batch if maxPartitionsPerRecoveryMessage is set
+	if s.maxPartitionsPerRecoveryMessage > 0 {
+		tmpRecoveryDecls := RecoveryDecls
+		var msgs []*types.SignedMessage
+		var RecoveryBatches [][]miner.RecoveryDeclaration
+		I := 0
+
+		// Create batched
+		for len(tmpRecoveryDecls) > s.maxPartitionsPerPostMessage {
+			Batch := tmpRecoveryDecls[len(tmpRecoveryDecls)-s.maxPartitionsPerRecoveryMessage:]
+			tmpRecoveryDecls = tmpRecoveryDecls[:len(tmpRecoveryDecls)-s.maxPartitionsPerPostMessage]
+			RecoveryBatches = append(RecoveryBatches, Batch)
+			I++
+		}
+
+		// Add remaining as new batch
+		RecoveryBatches = append(RecoveryBatches, tmpRecoveryDecls)
+
+		for _, Batch := range RecoveryBatches {
+			params := &miner.DeclareFaultsRecoveredParams{
+				Recoveries: Batch,
+			}
+
+			enc, aerr := actors.SerializeParams(params)
+
+			if aerr != nil {
+				return nil, xerrors.Errorf("could not serialize declare recoveries parameters: %w", aerr)
+			}
+
+			msg := &types.Message{
+				To:     s.actor,
+				Method: builtin.MethodsMiner.DeclareFaultsRecovered,
+				Params: enc,
+				Value:  types.NewInt(0),
+			}
+
+			spec := &api.MessageSendSpec{MaxFee: abi.TokenAmount(s.feeCfg.MaxWindowPoStGasFee)}
+
+			if err := s.prepareMessage(ctx, msg, spec); err != nil {
+				return nil, err
+			}
+
+			sm, err := s.api.MpoolPushMessage(ctx, msg, &api.MessageSendSpec{MaxFee: abi.TokenAmount(s.feeCfg.MaxWindowPoStGasFee)})
+
+			if err != nil {
+				return nil, xerrors.Errorf("pushing message to mpool: %w", err)
+			}
+
+			msgs = append(msgs, sm)
+		}
+
+		for _, v := range msgs {
+			mcid := v.Cid()
+			mcids = append(mcids, mcid)
+		}
+
+		return mcids, nil
+
+	}
+
+	params := &miner.DeclareFaultsRecoveredParams{
+		Recoveries: RecoveryDecls,
+	}
+
+	enc, aerr := actors.SerializeParams(params)
+	if aerr != nil {
+		return nil, xerrors.Errorf("could not serialize declare recoveries parameters: %w", aerr)
+	}
+
+	msg := &types.Message{
+		To:     s.actor,
+		Method: builtin.MethodsMiner.DeclareFaultsRecovered,
+		Params: enc,
+		Value:  types.NewInt(0),
+	}
+	spec := &api.MessageSendSpec{MaxFee: abi.TokenAmount(s.feeCfg.MaxWindowPoStGasFee)}
+	if err := s.prepareMessage(ctx, msg, spec); err != nil {
+		return nil, err
+	}
+	sm, err := s.api.MpoolPushMessage(ctx, msg, &api.MessageSendSpec{MaxFee: abi.TokenAmount(s.feeCfg.MaxWindowPoStGasFee)})
+	if err != nil {
+		return nil, xerrors.Errorf("pushing message to mpool: %w", err)
+	}
+
+	mcids = append(mcids, sm.Cid())
+	return mcids, nil
```

Comment:
This is mostly duplicated from lines 418-447. We should extract this core logic and just call it from both places.